// scichart/Charting/Model/IDataSeries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EDataSeriesType = void 0;
/**
 * Defines {@link BaseDataSeries | DataSeries} types available within SciChart's
 * {@link https://www.scichart.com/javascript-chart-features | JavaScript Charts}
 */
var EDataSeriesType;
(function (EDataSeriesType) {
    /**
     * Defines an {@link XyDataSeries}
     */
    EDataSeriesType["Xy"] = "Xy";
    /**
     * Defines an {@link XyyDataSeries}, with two Y-points for every X-Value
     */
    EDataSeriesType["Xyy"] = "Xyy";
    /**
     * Defines an {@link XyzDataSeries}
     */
    EDataSeriesType["Xyz"] = "Xyz";
    /**
     * Defines an {@link OhlcDataSeries} for
     * {@link https://www.scichart.com/javascript-chart-features | JavaScript Stock Charts} or
     * financial charts
     */
    EDataSeriesType["Ohlc"] = "Ohlc";
    /**
     * Defines a {@link UniformHeatmapDataSeries | DataSeries}
     */
    EDataSeriesType["HeatmapUniform"] = "UniformHeatmap";
})(EDataSeriesType = exports.EDataSeriesType || (exports.EDataSeriesType = {}));
