// scichart/utils/date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatUnixDateToHumanStringHHMM = exports.formatUnixDateToHumanStringDDMM = exports.formatUnixDateToHumanString = void 0;
/**
 * Result 11/23/2018
 * @param unixTimestamp
 */
exports.formatUnixDateToHumanString = (unixTimestamp) => {
    const res = new Date(unixTimestamp * 1000).toLocaleDateString("en-GB", {
        month: "numeric",
        year: "numeric",
        day: "numeric"
    });
    if (res === "Invalid Date") {
        return "";
    }
    return res;
};
exports.formatUnixDateToHumanStringDDMM = (unixTimestamp) => {
    const res = new Date(unixTimestamp * 1000).toLocaleDateString("en-GB", {
        day: "numeric",
        month: "numeric"
    });
    if (res === "Invalid Date") {
        return "";
    }
    return res;
};
exports.formatUnixDateToHumanStringHHMM = (unixTimestamp) => {
    const date = new Date(unixTimestamp * 1000);
    const hours = date.getUTCHours();
    const minutes = date.getUTCMinutes();
    if (isNaN(hours) || isNaN(minutes)) {
        return "";
    }
    const hoursString = hours <= 9 ? `0${hours}` : hours.toString(10);
    const minutesString = minutes <= 9 ? `0${minutes}` : minutes.toString(10);
    return `${hoursString}:${minutesString}`;
};
