// scichart/Charting3D/Visuals/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["CAMERA_POSITION"] = "CAMERA_POSITION";
    PROPERTY["CAMERA_TARGET"] = "CAMERA_TARGET";
    PROPERTY["CAMERA_ORTHOHEIGHT"] = "CAMERA_ORTHOHEIGHT";
    PROPERTY["CAMERA_ORTHOWIDTH"] = "CAMERA_ORTHOWIDTH";
    PROPERTY["CAMERA_FIELDOFVIEW"] = "CAMERA_FIELDOFVIEW";
    PROPERTY["CAMERA_PROJECTIONMODE"] = "CAMERA_PROJECTIONMODE";
    PROPERTY["CAMERA_ISATTACHED"] = "CAMERA_ISATTACHED";
    PROPERTY["CAMERA_ORBITALPITCH"] = "CAMERA_ORBITALPITCH";
    PROPERTY["CAMERA_ORBITALYAW"] = "CAMERA_ORBITALYAW";
    PROPERTY["CAMERA_RADIUS"] = "CAMERA_RADIUS";
    PROPERTY["CAMERA_FORWARD"] = "CAMERA_FORWARD";
    PROPERTY["CAMERA_SIDE"] = "CAMERA_SIDE";
    PROPERTY["CAMERA_UP"] = "CAMERA_UP";
    PROPERTY["CAMERA_UPGLOBAL"] = "CAMERA_UPGLOBAL";
    PROPERTY["CAMERA_ASPECTRATIO"] = "CAMERA_ASPECTRATIO";
    PROPERTY["CAMERA_ID"] = "CAMERA_ID";
    PROPERTY["CAMERA_FARCLIP"] = "CAMERA_FARCLIP";
    PROPERTY["CAMERA_NEARCLIP"] = "CAMERA_NEARCLIP";
    PROPERTY["SURFACE_CAMERA"] = "SURFACE_CAMERA";
    PROPERTY["SURFACE_XAXIS"] = "SURFACE_XAXIS";
    PROPERTY["SURFACE_YAXIS"] = "SURFACE_YAXIS";
    PROPERTY["SURFACE_ZAXIS"] = "SURFACE_ZAXIS";
    PROPERTY["SURFACE_VIEWPORT_MANAGER"] = "SURFACE_VIEWPORT_MANAGER";
    PROPERTY["SURFACE_WORLD_DIMENSIONS"] = "SURFACE_WORLD_DIMENSIONS";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
