// scichart/Charting/Visuals/loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Color_1 = require("../../types/Color");
const sciChartLoaderStylesId = "scichart_loader_styles_id";
const addLoaderStyles = () => {
    if (document.querySelector(`#${sciChartLoaderStylesId}`)) {
        return;
    }
    const head = document.head;
    const style = document.createElement("style");
    style.id = sciChartLoaderStylesId;
    head.appendChild(style);
    style.appendChild(document.createTextNode(loaderCss));
};
const addChartLoader = (domDivContainer) => {
    domDivContainer.style.backgroundColor = Color_1.EColor.BackgroundColor;
    const loaderDiv = document.createElement("div");
    loaderDiv.innerHTML = `<div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div>`;
    loaderDiv.classList.add("scichart_loader");
    domDivContainer.appendChild(loaderDiv);
    return loaderDiv;
};
const removeChartLoader = (domDivContainer, loaderDiv) => {
    domDivContainer.removeChild(loaderDiv);
    domDivContainer.style.backgroundColor = "transparent";
};
const loaderCss = `
    .scichart_loader {
      display: inline-block;
      position: relative;
      width: 80px;
      height: 80px;
      top: 50%;
      transform: translateY(-50%);
    }
    .scichart_loader div {
      animation: scichart_loader 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
      transform-origin: 40px 40px;
    }
    .scichart_loader div:after {
      content: " ";
      display: block;
      position: absolute;
      width: 7px;
      height: 7px;
      border-radius: 50%;
      background: #fff;
      margin: -4px 0 0 -4px;
    }
    .scichart_loader div:nth-child(1) {
      animation-delay: -0.036s;
    }
    .scichart_loader div:nth-child(1):after {
      top: 63px;
      left: 63px;
    }
    .scichart_loader div:nth-child(2) {
      animation-delay: -0.072s;
    }
    .scichart_loader div:nth-child(2):after {
      top: 68px;
      left: 56px;
    }
    .scichart_loader div:nth-child(3) {
      animation-delay: -0.108s;
    }
    .scichart_loader div:nth-child(3):after {
      top: 71px;
      left: 48px;
    }
    .scichart_loader div:nth-child(4) {
      animation-delay: -0.144s;
    }
    .scichart_loader div:nth-child(4):after {
      top: 72px;
      left: 40px;
    }
    .scichart_loader div:nth-child(5) {
      animation-delay: -0.18s;
    }
    .scichart_loader div:nth-child(5):after {
      top: 71px;
      left: 32px;
    }
    .scichart_loader div:nth-child(6) {
      animation-delay: -0.216s;
    }
    .scichart_loader div:nth-child(6):after {
      top: 68px;
      left: 24px;
    }
    .scichart_loader div:nth-child(7) {
      animation-delay: -0.252s;
    }
    .scichart_loader div:nth-child(7):after {
      top: 63px;
      left: 17px;
    }
    .scichart_loader div:nth-child(8) {
      animation-delay: -0.288s;
    }
    .scichart_loader div:nth-child(8):after {
      top: 56px;
      left: 12px;
    }
    @keyframes scichart_loader {
      0% {
        transform: rotate(0deg);
      }
      100% {
        transform: rotate(360deg);
      }
    }
`;
const loader = {
    addLoaderStyles,
    addChartLoader,
    removeChartLoader
};
exports.default = loader;
