// scichart/Charting3D/Visuals/Axis/AxisBase3DLabelStyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxisBase3DLabelStyle = void 0;
const constants_1 = require("./constants");
class AxisBase3DLabelStyle {
    constructor(parentAxis) {
        this.dpiScalingProperty = 96;
        this.fontFamilyProperty = "Arial";
        this.fontSizeProperty = 14;
        this.foregroundProperty = "#c8c7c3";
        this.parentAxis = parentAxis;
    }
    get alignment() {
        return this.alignmentProperty;
    }
    set alignment(value) {
        this.alignmentProperty = value;
        this.notifyPropertyChanged(constants_1.TEXT_STYLE_PROPERTY.ALIGNMENT);
    }
    get dpiScaling() {
        return this.dpiScalingProperty;
    }
    set dpiScaling(value) {
        this.dpiScalingProperty = value;
        this.notifyPropertyChanged(constants_1.TEXT_STYLE_PROPERTY.DPI_SCALING);
    }
    get fontFamily() {
        return this.fontFamilyProperty;
    }
    set fontFamily(value) {
        this.fontFamilyProperty = value;
        this.notifyPropertyChanged(constants_1.TEXT_STYLE_PROPERTY.FONT_FAMILY);
    }
    get fontSize() {
        return this.fontSizeProperty;
    }
    set fontSize(value) {
        this.fontSizeProperty = value;
        this.notifyPropertyChanged(constants_1.TEXT_STYLE_PROPERTY.FONT_SIZE);
    }
    get foreground() {
        return this.foregroundProperty;
    }
    set foreground(value) {
        this.foregroundProperty = value;
        this.notifyPropertyChanged(constants_1.TEXT_STYLE_PROPERTY.FOREGROUND);
    }
    notifyPropertyChanged(propertyName) {
        var _a;
        const invalidateParentCallback = (_a = this.parentAxis) === null || _a === void 0 ? void 0 : _a.invalidateParentCallback;
        if (invalidateParentCallback) {
            invalidateParentCallback();
        }
    }
}
exports.AxisBase3DLabelStyle = AxisBase3DLabelStyle;
