// scichart/Charting3D/Visuals/Axis/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEXT_STYLE_PROPERTY = exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["AXIS_PLANE_BACKGROUND_FILL"] = "AXIS_PLANE_BACKGROUND_FILL";
    PROPERTY["IS_Y_AXIS"] = "IS_Y_AXIS";
    PROPERTY["IS_Z_AXIS"] = "IS_Z_AXIS";
    PROPERTY["LABEL_DEPTH_TEST"] = "LABEL_DEPTH_TEST";
    PROPERTY["NEGATIVE_SIDE_CLIPPING"] = "NEGATIVE_SIDE_CLIPPING";
    PROPERTY["PLANE_BORDER_COLOR"] = "PLANE_BORDER_COLOR";
    PROPERTY["PLANE_BORDER_THICKNESS"] = "PLANE_BORDER_THICKNESS";
    PROPERTY["POSITIVE_SIDE_CLIPPING"] = "POSITIVE_SIDE_CLIPPING";
    PROPERTY["TICK_LABEL_ALIGNMENT"] = "TICK_LABEL_ALIGNMENT";
    PROPERTY["TICK_LABELS_OFFSET"] = "TICK_LABELS_OFFSET";
    PROPERTY["TITLE_OFFSET"] = "TITLE_OFFSET";
    PROPERTY["BACKGROUND_COLOR"] = "BACKGROUND_COLOR";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
var TEXT_STYLE_PROPERTY;
(function (TEXT_STYLE_PROPERTY) {
    TEXT_STYLE_PROPERTY["ALIGNMENT"] = "ALIGNMENT";
    TEXT_STYLE_PROPERTY["DPI_SCALING"] = "DPI_SCALING";
    TEXT_STYLE_PROPERTY["FONT_FAMILY"] = "FONT_FAMILY";
    TEXT_STYLE_PROPERTY["FONT_SIZE"] = "FONT_SIZE";
    TEXT_STYLE_PROPERTY["FOREGROUND"] = "FOREGROUND";
})(TEXT_STYLE_PROPERTY = exports.TEXT_STYLE_PROPERTY || (exports.TEXT_STYLE_PROPERTY = {}));
