// scichart/Core/Dictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dictionary = void 0;
/**
 * Generic dictionary class keyed by string
 */
class Dictionary {
    constructor() {
        this.items = {};
        this.countProperty = 0;
    }
    /**
     * @inheritDoc
     */
    containsKey(key) {
        return this.items.hasOwnProperty(key);
    }
    /**
     * @inheritDoc
     */
    get count() {
        return this.countProperty;
    }
    /**
     * @inheritDoc
     */
    add(key, value) {
        if (!this.items.hasOwnProperty(key)) {
            this.countProperty++;
        }
        this.items[key] = value;
    }
    /**
     * @inheritDoc
     */
    remove(key) {
        const val = this.items[key];
        delete this.items[key];
        this.countProperty--;
        return val;
    }
    /**
     * @inheritDoc
     */
    item(key) {
        return this.items[key];
    }
    /**
     * @inheritDoc
     */
    get keys() {
        return Object.keys(this.items);
    }
    /**
     * @inheritDoc
     */
    get values() {
        return this.keys.map(key => this.items[key]);
    }
    /**
     * Clears the dictionary
     */
    clear() {
        this.items = {};
        this.countProperty = 0;
    }
}
exports.Dictionary = Dictionary;
