// scichart/Charting/Themes/defaultThemeColors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultThemeColors = void 0;
exports.defaultThemeColors = {
    sciChartBackground: "#1C1C1EFF",
    gridBorderBrush: "#5A5B5BFF",
    axisBandsFill: "#202123E1",
    tickTextBrush: "#A6A7ACFF",
    majorGridLineBrush: "#323539FF",
    minorGridLineBrush: "#232426FF",
    gridBackgroundBrush: "TRANSPARENT",
    rolloverLineBrush: "#42b64933",
    cursorLineBrush: "#228B22FF",
    rubberBandFillBrush: "#42b64933",
    rubberBandStrokeBrush: "#42b64977",
    legendBackgroundBrush: "#1D2C35FF",
    labelBackgroundBrush: "#42b649AA",
    labelBorderBrush: "#42b649FF",
    labelForegroundBrush: "#EEEEEEFF",
    textAnnotationForeground: "#EEEEEEFF",
    textAnnotationBackground: "#42b649AA",
    annotationsGripsBorderBrush: "#EEEEEEFF",
    annotationsGripsBackroundBrush: "#FFFFFF33",
    overviewFillBrush: "#262728BB",
    scrollbarBackgroundBrush: "#262728FF",
    scrollbarBorderBrush: "#121212FF",
    scrollbarGripsBackgroundBrush: "#535353FF",
    scrollbarViewportBackgroundBrush: "#222222FF",
    scrollbarViewportBorderBrush: "#232323FF",
    upWickColor: "#50ff50FF",
    downWickColor: "#ff5050FF",
    upBodyBrush: "#50ff50B2",
    downBodyBrush: "#ff5050B2",
    upBandSeriesLineColor: "#279B27FF",
    downBandSeriesLineColor: "#FF1919FF",
    upBandSeriesFillColor: "#279B2733",
    downBandSeriesFillColor: "#FF191933",
    mountainAreaBrush: "#4083B777",
    mountainLineColor: "#C6E6FFFF",
    lineSeriesColor: "#C6E6FFFF",
    columnLineColor: "#FFFFFFFF",
    columnFillBrush: "#FFFFFFFF",
    defaultColorMapBrush: {
        point_0: "DARKBLUE",
        point_0_5: "CORNFLOWERBLUE",
        point_1: "#FF22AAFF"
    },
    shadowEffectColor: "#000000FF",
    planeBorderColor: "#333333FF",
    axisPlaneBackgroundFill: "TRANSPARENT",
    axis3DBandsFill: "#202123E1",
    axisBorder: "#00000000",
    axisTitleColor: "#C8C7C3FF"
};
