// scichart/Charting/Themes/SciChartJSLightTheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SciChartJSLightTheme = void 0;
/**
 * An implementation of {@link IThemeProvider} which provides a light theme
 * @decription Applied to a 2D {@link SciChartSurface}, or a 3D {@link SciChart3DSurface}, the ThemeProvider may be
 * applied using the {@link SciChartSurface.applyTheme | applyTheme} method, where it is passed down to child components
 * so that all children of the chart get the same theme.
 *
 * For example:
 * ```ts
 * // Applying a built-in dark theme
 * const sciChartSurface: SciChartSurface;
 * sciChartSurface.applyTheme(new SciChartJSDarkTheme());
 * // Or light theme
 * sciChartSurface.applyTheme(new SciChartJSLightTheme());
 *
 * // Applying a custom theme
 * export class MyCustomTheme implements IThemeProvider {
 *    // todo: implement IThemeProvider interface and apply properties
 * }
 *
 * sciChartSurface.applyTheme(new MyCustomTheme());
 * ```
 */
class SciChartJSLightTheme {
    constructor() {
        /** @inheritDoc */
        this.sciChartBackground = "#F9F9F9FF";
        /** @inheritDoc */
        this.gridBorderBrush = "#33333399";
        /** @inheritDoc */
        this.axisBandsFill = "#AAAAAA11";
        /** @inheritDoc */
        this.tickTextBrush = "#333333FF";
        /** @inheritDoc */
        this.majorGridLineBrush = "#AAAAAAFF";
        /** @inheritDoc */
        this.minorGridLineBrush = "#CFCFCFFF";
        /** @inheritDoc */
        this.gridBackgroundBrush = "#05333377";
        /** @inheritDoc */
        this.rolloverLineBrush = "#33333333";
        /** @inheritDoc */
        this.cursorLineBrush = "#33333355";
        /** @inheritDoc */
        this.rubberBandFillBrush = "#33333333";
        /** @inheritDoc */
        this.rubberBandStrokeBrush = "#33333377";
        /** @inheritDoc */
        this.legendBackgroundBrush = "#33333333";
        /** @inheritDoc */
        this.labelBackgroundBrush = "#D0D0D0BB";
        /** @inheritDoc */
        this.labelBorderBrush = "#33333377";
        /** @inheritDoc */
        this.labelForegroundBrush = "#555555FF";
        /** @inheritDoc */
        this.textAnnotationForeground = "#000000FF";
        /** @inheritDoc */
        this.textAnnotationBackground = "#FFFFFFFF";
        /** @inheritDoc */
        this.annotationsGripsBorderBrush = "#232323FF";
        /** @inheritDoc */
        this.annotationsGripsBackroundBrush = "#FFFFFF33";
        /** @inheritDoc */
        this.overviewFillBrush = "#33333322";
        /** @inheritDoc */
        this.scrollbarBackgroundBrush = "#33333322";
        /** @inheritDoc */
        this.scrollbarBorderBrush = "#12121255";
        /** @inheritDoc */
        this.scrollbarGripsBackgroundBrush = "#FFFFFF66";
        /** @inheritDoc */
        this.scrollbarViewportBackgroundBrush = "#FFFFFF44";
        /** @inheritDoc */
        this.scrollbarViewportBorderBrush = "#12121255";
        /** @inheritDoc */
        this.upWickColor = "#52CC54FF";
        /** @inheritDoc */
        this.downWickColor = "#E26565FF";
        /** @inheritDoc */
        this.upBodyBrush = "#52CC54A0";
        /** @inheritDoc */
        this.downBodyBrush = "#E26565D0";
        /** @inheritDoc */
        this.upBandSeriesLineColor = "#52CC54FF";
        /** @inheritDoc */
        this.downBandSeriesLineColor = "#E26565FF";
        /** @inheritDoc */
        this.upBandSeriesFillColor = "#52CC5490";
        /** @inheritDoc */
        this.downBandSeriesFillColor = "#E26565A0";
        /** @inheritDoc */
        this.mountainAreaBrush = "#76B7E2B4";
        /** @inheritDoc */
        this.mountainLineColor = "#777777FF";
        /** @inheritDoc */
        this.lineSeriesColor = "#777777FF";
        /** @inheritDoc */
        this.columnLineColor = "#777777FF";
        /** @inheritDoc */
        this.columnFillBrush = "#777777FF";
        /** @inheritDoc */
        this.defaultColorMapBrush = [
            { offset: 0, color: "DARKBLUE" },
            { offset: 0.5, color: "CORNFLOWERBLUE" },
            { offset: 1, color: "#FF22AAFF" }
        ];
        /** @inheritDoc */
        this.axisTitleColor = "#777777FF";
        /** @inheritDoc */
        this.shadowEffectColor = "#A0AABAFA";
        /** @inheritDoc */
        this.planeBorderColor = "#EEEEEEFF";
        /** @inheritDoc */
        this.axisPlaneBackgroundFill = "TRANSPARENT";
        /** @inheritDoc */
        this.axis3DBandsFill = "#33333333";
    }
}
exports.SciChartJSLightTheme = SciChartJSLightTheme;
