// scichart/Charting/LayoutManager/ChartLayoutState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChartLayoutState = void 0;
/**
 * @ignore
 */
class ChartLayoutState {
    constructor() {
        // TODO: add inner if needed
        // public leftInnerAreaSize: number = ChartLayoutState.defaultSize;
        // public topInnerAreaSize: number = ChartLayoutState.defaultSize;
        // public bottomInnerAreaSize: number = ChartLayoutState.defaultSize;
        // public rightInnerAreaSize: number = ChartLayoutState.defaultSize;
        this.leftOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.topOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.bottomOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.rightOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
    }
    clear() {
        this.leftOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.topOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.bottomOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
        this.rightOuterAreaSize = ChartLayoutState.DEFAULT_SIZE;
    }
}
exports.ChartLayoutState = ChartLayoutState;
ChartLayoutState.DEFAULT_SIZE = 0;
