// scichart/Charting/Visuals/Annotations/AdornerLayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdornerLayer = void 0;
class AdornerLayer {
    constructor(scs) {
        this.parentSurface = scs;
    }
    selectAnnotation(args) {
        const scs = this.parentSurface;
        let selectedAnnotation;
        scs.annotations.asArray().forEach(an => {
            const anBase = an;
            if (!selectedAnnotation && anBase.isEditable) {
                const isAnnotationSelected = anBase.isPointWithinBounds(args);
                if (isAnnotationSelected) {
                    selectedAnnotation = anBase;
                }
            }
            else {
                anBase.isSelected = false;
            }
        });
        this.selectedAnnotationProperty = selectedAnnotation;
        return !!selectedAnnotation;
    }
    get selectedAnnotation() {
        return this.selectedAnnotationProperty;
    }
    get isAnnotationSelected() {
        return !!this.selectedAnnotationProperty;
    }
}
exports.AdornerLayer = AdornerLayer;
