// scichart/Charting/Visuals/Annotations/IAnnotation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EAnnotationType = exports.EAnnotationLayer = void 0;
/**
 * Defines the layer where {@link IAnnotation | Annotations} are drawn
 * when added to the {@link SciChartSurface.annotations} collection
 */
var EAnnotationLayer;
(function (EAnnotationLayer) {
    /**
     * The {@link IAnnotation | Annotation} will be displayed above the chart series and grid
     * @remarks
     * This is the default value for {@link IAnnotation | Annotations}
     */
    EAnnotationLayer[EAnnotationLayer["AboveChart"] = 0] = "AboveChart";
    /**
     * The {@link IAnnotation | Annotation} will be displayed below the chart series and grid.
     * @remarks
     * Use this for watermarks, e.g. showing an image or text behind the chart.
     * The Grid lines and Axis Bands will show over the annotation,
     * so consider setting {@link AxisCore.axisBandsFill} to a semi-transparent color to avoid this.
     */
    EAnnotationLayer[EAnnotationLayer["BelowChart"] = 1] = "BelowChart";
})(EAnnotationLayer = exports.EAnnotationLayer || (exports.EAnnotationLayer = {}));
/**
 * Defines the type of {@link IAnnotation | Annotation}.
 * Annotations are overlays or markers added to the {@link SciChartSurface.annotations} collection
 */
var EAnnotationType;
(function (EAnnotationType) {
    /**
     * The {@link IAnnotation | Annotation} is a {@link RenderContextAnnotationBase | RenderContextAnnotation}
     * @remarks
     * {@link RenderContextAnnotationBase | RenderContextAnnotations} are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContext"] = 0] = "RenderContext";
    /**
     * The Annotation is an {@link SvgAnnotation | SvgAnnotation}
     * @remarks
     * {@link SvgAnnotation | SvgAnnotations} are drawn using SVG where available.
     * This allows for more flexible annotations over the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript Chart},
     * such as text or other custom shapes
     */
    EAnnotationType[EAnnotationType["SVG"] = 1] = "SVG";
    /**
     * The Annotation is an {@link BoxAnnotation | BoxAnnotation}
     * @remarks
     * {@link BoxAnnotation | BoxAnnotation} * are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContextBoxAnnotation"] = 2] = "RenderContextBoxAnnotation";
    /**
     * The Annotation is an {@link LineAnnotation | LineAnnotation}
     * @remarks
     * {@link LineAnnotation | BoxAnnotation} * are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContextLineAnnotation"] = 3] = "RenderContextLineAnnotation";
    /**
     * The Annotation is an {@link HorizontalLineAnnotation | HorizontalLineAnnotation}
     * @remarks
     * {@link HorizontalLineAnnotation | HorizontalLineAnnotation} * are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContextHorizontalLineAnnotation"] = 4] = "RenderContextHorizontalLineAnnotation";
    /**
     * The Annotation is an {@link VerticalLineAnnotation | VerticalLineAnnotation}
     * @remarks
     * {@link VerticalLineAnnotation | VerticalLineAnnotation} * are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContextVerticalLineAnnotation"] = 5] = "RenderContextVerticalLineAnnotation";
    /**
     * The Annotation is an {@link AxisMarkerAnnotation | AxisMarkerAnnotation}
     * @remarks
     * {@link AxisMarkerAnnotation | AxisMarkerAnnotation} * are drawn on the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript chart} using WebGL2 where available.
     * They differ from {@link SvgAnnotation | SvgAnnotations} which use slower,
     * but more flexible SVG to draw annotations on the chart
     */
    EAnnotationType[EAnnotationType["RenderContextAxisMarkerAnnotation"] = 6] = "RenderContextAxisMarkerAnnotation";
    /**
     * The Annotation is an {@link TextAnnotation | TextAnnotation}
     * @remarks
     * {@link SvgAnnotation | SvgAnnotations} are drawn using SVG where available.
     * This allows for more flexible annotations over the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript Chart},
     * such as text or other custom shapes
     */
    EAnnotationType[EAnnotationType["SVGTextAnnotation"] = 7] = "SVGTextAnnotation";
    /**
     * The Annotation is an {@link SVGCustomAnnotation | SVGCustomAnnotation}
     * @remarks
     * {@link SvgAnnotation | SvgAnnotations} are drawn using SVG where available.
     * This allows for more flexible annotations over the
     * {@link https://www.scichart.com/javascript-chart-features | Javascript Chart},
     * such as text or other custom shapes
     */
    EAnnotationType[EAnnotationType["SVGCustomAnnotation"] = 8] = "SVGCustomAnnotation";
})(EAnnotationType = exports.EAnnotationType || (exports.EAnnotationType = {}));
