// scichart/Charting/Drawing/BaseCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCache = void 0;
const Deleter_1 = require("../../Core/Deleter");
const WebGlRenderContext2D_1 = require("./WebGlRenderContext2D");
/**
 * @ignore
 */
class BaseCache {
    /**
     * Creates an instance of {@link BaseCache}
     * @param webAssemblyContext the {@link TSciChart | SciChart WebAssembly Context} containing native methods
     * and access to our underlying WebGL2 rendering engine
     */
    constructor(webAssemblyContext) {
        this.webAssemblyContext = webAssemblyContext;
        if (webAssemblyContext) {
            // add reference of the current instance to global collection of cached resources
            WebGlRenderContext2D_1.WebGlRenderContext2D.webGlResourcesRefs.add(this);
        }
    }
    /**
     * Retrieves the cached value
     */
    get value() {
        return this.cachedEntity;
    }
    /**
     * @inheritDoc
     */
    invalidateCache() {
        this.cachedEntity = Deleter_1.deleteSafe(this.cachedEntity);
    }
    /**
     * @inheritDoc
     */
    resetCache() {
        this.invalidateCache();
    }
    /**
     * @inheritDoc
     */
    delete() {
        this.cachedEntity = Deleter_1.deleteSafe(this.cachedEntity);
        if (this.webAssemblyContext) {
            // remove reference of the current instance to global collection of cached resources
            WebGlRenderContext2D_1.WebGlRenderContext2D.webGlResourcesRefs.delete(this);
        }
    }
}
exports.BaseCache = BaseCache;
