// scichart/Charting/Drawing/SolidBrushCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolidBrushCache = void 0;
const parseColor_1 = require("../../utils/parseColor");
// TODO: this class can be removed and its usage replaced by BrushCache class
/**
 * @ignore
 */
class SolidBrushCache {
    constructor(webAssemblyContext) {
        this.webAssemblyContext = webAssemblyContext;
    }
    /**
     * Creates or fetches a new Brush with the specified color string and properties
     * @param color
     * @param transparent
     */
    newBrush(color, transparent = false) {
        // Return brush from cache
        if (this.brush && color === this.color && transparent === this.transparent) {
            return this.brush;
        }
        // Create new brush
        if (this.brush) {
            this.brush.delete();
        }
        this.color = color;
        this.transparent = transparent;
        this.brush = new this.webAssemblyContext.SCRTSolidBrush(parseColor_1.parseColorToUIntArgb(color), transparent);
        return this.brush;
    }
}
exports.SolidBrushCache = SolidBrushCache;
