// scichart/Charting/Drawing/WebGlPen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebGlPen = void 0;
const Deleter_1 = require("../../Core/Deleter");
const colorUtil_1 = require("../../utils/colorUtil");
const constants_1 = require("./constants");
/**
 * The WebGlPen is a pen for polygon stroke, line strokes, which can be passed to SciChart's WebGL / WebAssembly graphics engine
 */
class WebGlPen {
    /**
     * Creates an instance of WebGlPen
     * @param scrtPen the inner {@link SCRTPen} which can be passed to SciChart's WebAssembly WebGL engine
     */
    constructor(scrtPen) {
        this.scrtPenProperty = scrtPen;
        this.originalColor = scrtPen === null || scrtPen === void 0 ? void 0 : scrtPen.m_uiColor;
    }
    /**
     * the inner {@link SCRTPen} which can be passed to SciChart's WebAssembly WebGL engine
     */
    get scrtPen() {
        return this.scrtPenProperty;
    }
    /**
     * @inheritDoc
     */
    delete() {
        this.scrtPenProperty = Deleter_1.deleteSafe(this.scrtPenProperty);
    }
    /**
     * @inheritDoc
     */
    getPenType() {
        return constants_1.EDrawingTypes.WasmPen;
    }
    /**
     * @inheritDoc
     */
    setOpacity(opacity) {
        this.scrtPen.m_uiColor = colorUtil_1.uintArgbColorMultiplyOpacity(this.originalColor, opacity);
    }
}
exports.WebGlPen = WebGlPen;
