// scichart/Charting/Drawing/WebGlBrush.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebGlBrush = void 0;
const Deleter_1 = require("../../Core/Deleter");
const constants_1 = require("./constants");
/**
 * The WebGLBrush is a brush for polygon fills, rectangle fills, which can be passed to SciChart's WebGL / WebAssembly graphics engine
 */
class WebGlBrush {
    /**
     * Creates an instance of WebGlBrush
     * @param scrtBrush the inner {@link SCRTBrush} which can be passed to SciChart's WebAssembly WebGL engine
     */
    constructor(scrtBrush) {
        this.scrtBrushProperty = scrtBrush;
    }
    /**
     * Get the inner {@link SCRTBrush} which can be passed to SciChart's WebAssembly WebGL engine
     */
    get scrtBrush() {
        return this.scrtBrushProperty;
    }
    /**
     * @inheritDoc
     */
    delete() {
        this.scrtBrushProperty = Deleter_1.deleteSafe(this.scrtBrushProperty);
    }
    /**
     * @inheritDoc
     */
    getBrushType() {
        return constants_1.EDrawingTypes.WasmBrush;
    }
    /**
     * @inheritDoc
     */
    setOpacity(opacity) {
        this.scrtBrushProperty.SetOpacity(opacity);
    }
}
exports.WebGlBrush = WebGlBrush;
