// scichart/Charting3D/Visuals/RenderableSeries/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["CELL_HARDNESS_FACTOR"] = "CELL_HARDNESS_FACTOR";
    PROPERTY["COLOR_MAP_TEXTURE_SIZE"] = "COLOR_MAP_TEXTURE_SIZE";
    PROPERTY["CONTOUR_STROKE"] = "CONTOUR_STROKE";
    PROPERTY["CONTOUR_OFFSET"] = "CONTOUR_OFFSET";
    PROPERTY["CONTOUR_INTERVAL"] = "CONTOUR_INTERVAL";
    PROPERTY["CONTOUR_STROKE_THICKNESS"] = "CONTOUR_STROKE_THICKNESS";
    PROPERTY["DATA_SERIES"] = "DATA_SERIES";
    PROPERTY["DRAW_MESH_AS"] = "DRAW_MESH_AS";
    PROPERTY["DRAW_SKIRT"] = "DRAW_SKIRT";
    PROPERTY["FILL"] = "FILL";
    PROPERTY["GRADIENT_STOPS"] = "GRADIENT_STOPS";
    PROPERTY["HEIGHT_SCALE_FACTOR"] = "HEIGHT_SCALE_FACTOR";
    PROPERTY["HIGHLIGHT"] = "HIGHLIGHT";
    PROPERTY["IS_VISIBLE"] = "IS_VISIBLE";
    PROPERTY["LIGHTING_FACTOR"] = "LIGHTING_FACTOR";
    PROPERTY["MAXIMUM"] = "MAXIMUM";
    PROPERTY["MESH_COLOR_PALETTE"] = "MESH_COLOR_PALETTE";
    PROPERTY["MESH_PALETTE_MODE"] = "MESH_PALETTE_MODE";
    PROPERTY["MESH_RESOLUTION"] = "MESH_RESOLUTION";
    PROPERTY["MINIMUM"] = "MINIMUM";
    PROPERTY["OPACITY"] = "OPACITY";
    PROPERTY["PALETTE_PROVIDER"] = "PALETTE_PROVIDER";
    PROPERTY["PARENT_SURFACE"] = "PARENT_SURFACE";
    PROPERTY["POINT_MARKER3D"] = "POINT_MARKER3D";
    PROPERTY["SCENE_ENTITY"] = "SCENE_ENTITY";
    PROPERTY["SHININESS"] = "SHININESS";
    PROPERTY["STROKE"] = "STROKE";
    PROPERTY["STROKE_THICKNESS"] = "STROKE_THICKNESS";
    PROPERTY["Y_OFFSET"] = "Y_OFFSET";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
