// scichart/Charting/Visuals/RenderableSeries/SplineMountainRenderableSeries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplineMountainRenderableSeries = void 0;
const Deleter_1 = require("../../../Core/Deleter");
const AxisType_1 = require("../../../types/AxisType");
const SeriesType_1 = require("../../../types/SeriesType");
const animationHelper_1 = require("./Animations/animationHelper");
const BaseMountainRenderableSeries_1 = require("./BaseMountainRenderableSeries");
const constants_1 = require("./constants");
/**
 * Defines a mountain/area series or JavaScript mountain chart type in the SciChart's High Performance Real-time
 * {@link https://www.scichart.com/javascript-chart-features | JavaScript Charts}
 * @remarks
 * To add a spline mountain series to a {@link SciChartSurface} you need to declare both the {@link SplineMountainRenderableSeries | RenderableSeries}
 * and a {@link XyDataSeries | DataSeries}. Simplified code sample below:
 *
 * ```ts
 * const sciChartSurface: SciChartSurface;
 * const wasmContext: TSciChart;
 * // Create and fill the dataseries
 * const dataSeries = new XyDataSeries(wasmContext);
 * dataSeries.append(1,2);
 * dataSeries.append(1,2);
 * // Create the renderableSeries
 * const mountainSeries = new SplineMountainRenderableSeries(wasmContext);
 * mountainSeries.dataSeries = dataSeries;
 * // append to the SciChartSurface
 * sciChartSurface.renderableSeries.add(mountainSeries);
 * ```
 */
class SplineMountainRenderableSeries extends BaseMountainRenderableSeries_1.BaseMountainRenderableSeries {
    /**
     * Creates an instance of the {@link SplineMountainRenderableSeries}
     * @param webAssemblyContext The {@link TSciChart | SciChart WebAssembly Context} containing
     * native methods and access to our WebGL2 WebAssembly Drawing Engine
     * @param options optional parameters of type {@link ISplineMountainRenderableSeriesOptions} applied when constructing the series type
     */
    constructor(webAssemblyContext, options) {
        var _a, _b;
        super(webAssemblyContext, options);
        this.type = SeriesType_1.ESeriesType.SplineMountainSeries;
        this.recalculateSpline = true;
        this.interpolationPointsProperty = 0;
        this.xSplineValues = new webAssemblyContext.DoubleVector();
        this.ySplineValues = new webAssemblyContext.DoubleVector();
        this.interpolationPointsProperty = (_a = options === null || options === void 0 ? void 0 : options.interpolationPoints) !== null && _a !== void 0 ? _a : this.interpolationPointsProperty;
        // This should be in the end after all DoubleVectors used for animation have been created
        this.animation = (_b = options === null || options === void 0 ? void 0 : options.animation) !== null && _b !== void 0 ? _b : this.animation;
    }
    /**
     * Gets or sets the interpolationPoints being used for the Spline
     */
    get interpolationPoints() {
        return this.interpolationPointsProperty;
    }
    /**
     * Gets or sets the interpolationPoints being used for the Spline
     */
    set interpolationPoints(value) {
        this.interpolationPointsProperty = value;
        this.notifyPropertyChanged(constants_1.PROPERTY.INTERPOLATION_POINTS);
    }
    /** @inheritDoc */
    delete() {
        this.xSplineValues = Deleter_1.deleteSafe(this.xSplineValues);
        this.ySplineValues = Deleter_1.deleteSafe(this.ySplineValues);
        super.delete();
    }
    /**
     * Updates spline values
     */
    updateSplineValues() {
        if (!(this.dataSeries && this.recalculateSpline && this.parentSurface)) {
            return;
        }
        const isCategoryAxis = this.xAxis.type === AxisType_1.EAxisType.CategoryAxis;
        const dataSeries = this.dataSeries;
        const xValues = isCategoryAxis ? dataSeries.getNativeIndexes() : dataSeries.getNativeXValues();
        const yValues = dataSeries.getNativeYValues();
        const initialSize = xValues.size();
        this.webAssemblyContext.SCRTSplineHelperCubicSpline(xValues, yValues, this.xSplineValues, this.ySplineValues, initialSize, this.interpolationPoints, dataSeries.containsNaN);
        this.recalculateSpline = false;
    }
    /**
     * @inheritDoc
     */
    onAnimate(progress) {
        var _a;
        super.onAnimate(progress);
        const dataSeries = this.dataSeries;
        if (!dataSeries) {
            (_a = this.animationToken) === null || _a === void 0 ? void 0 : _a.cancelAnimation();
        }
        else {
            this.updateSplineValues();
            animationHelper_1.calculateAnimationValues(this.webAssemblyContext, this.ySplineValues, this.yAnimationValues, this.animation, progress);
        }
    }
    /**
     * @inheritDoc
     */
    dataSeriesDataChanged() {
        this.recalculateSpline = true;
        super.dataSeriesDataChanged();
    }
}
exports.SplineMountainRenderableSeries = SplineMountainRenderableSeries;
