// scichart/Charting/Visuals/RenderableSeries/XyScatterRenderableSeries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyScatterRenderableSeries = void 0;
const SeriesType_1 = require("../../../types/SeriesType");
const animationHelper_1 = require("./Animations/animationHelper");
const BaseRenderableSeries_1 = require("./BaseRenderableSeries");
const PointMarkerDrawingProvider_1 = require("./DrawingProviders/PointMarkerDrawingProvider");
const ScatterSeriesHitTestProvider_1 = require("./HitTest/ScatterSeriesHitTestProvider");
/**
 * Defines a scatter-series or scatter chart type in the SciChart's High Performance Real-time
 * {@link https://www.scichart.com/javascript-chart-features | JavaScript Charts}
 * @remarks
 * To add a scatter series to a {@link SciChartSurface} you need to declare both the {@link XyScatterRenderableSeries | RenderableSeries}
 * and a {@link XyDataSeries | DataSeries}. Simplified code sample below:
 *
 * ```ts
 * const sciChartSurface: SciChartSurface;
 * const wasmContext: TSciChart;
 * // Create and fill the dataseries
 * const dataSeries = new XyDataSeries(wasmContext);
 * dataSeries.append(1,2);
 * dataSeries.append(1,2);
 * // Create the renderableSeries
 * const scatterSeries = new XyScatterRenderableSeries(wasmContext);
 * scatterSeries.dataSeries = dataSeries;
 * scatterSeries.pointMarker = new EllipsePointMarker(wasmContext, {
 *     width: 9,
 *     height: 9,
 *     fill: "#FF0000",
 *     stroke: "#0000FF",
 *     strokeThickness: 1
 * });
 * // append to the SciChartSurface
 * sciChartSurface.renderableSeries.add(scatterSeries);
 * ```
 */
class XyScatterRenderableSeries extends BaseRenderableSeries_1.BaseRenderableSeries {
    /**
     * Creates an instance of the {@link XyScatterRenderableSeries}
     * @param webAssemblyContext The {@link TSciChart | SciChart WebAssembly Context} containing
     * native methods and access to our WebGL2 WebAssembly Drawing Engine
     * @param options Optional parameters of type {@link IBaseRenderableSeriesOptions} to configure the series
     */
    constructor(webAssemblyContext, options) {
        var _a;
        super(webAssemblyContext, options);
        this.type = SeriesType_1.ESeriesType.ScatterSeries;
        this.drawingProviders = [];
        this.drawingProviders.push(new PointMarkerDrawingProvider_1.PointMarkerDrawingProvider(webAssemblyContext, this));
        // This should be in the end after all DoubleVectors used for animation have been created
        this.animation = (_a = options === null || options === void 0 ? void 0 : options.animation) !== null && _a !== void 0 ? _a : this.animation;
    }
    /**
     * @inheritDoc
     */
    newHitTestProvider() {
        return new ScatterSeriesHitTestProvider_1.ScatterSeriesHitTestProvider(this);
    }
    /**
     * @inheritDoc
     */
    onAnimate(progress) {
        var _a;
        super.onAnimate(progress);
        const dataSeries = this.dataSeries;
        if (!dataSeries) {
            (_a = this.animationToken) === null || _a === void 0 ? void 0 : _a.cancelAnimation();
        }
        else {
            animationHelper_1.calculateAnimationValues(this.webAssemblyContext, dataSeries.getNativeYValues(), this.yAnimationValues, this.animation, progress);
        }
    }
}
exports.XyScatterRenderableSeries = XyScatterRenderableSeries;
