// scichart/types/AxisAlignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIsVertical = exports.getIsHorizontal = exports.EAxisAlignment = void 0;
/**
 * Enumeration constants to define {@link AxisBase2D.alignment}
 */
var EAxisAlignment;
(function (EAxisAlignment) {
    EAxisAlignment["Right"] = "Right";
    EAxisAlignment["Left"] = "Left";
    EAxisAlignment["Top"] = "Top";
    EAxisAlignment["Bottom"] = "Bottom";
})(EAxisAlignment = exports.EAxisAlignment || (exports.EAxisAlignment = {}));
exports.getIsHorizontal = (alignment) => {
    if ([EAxisAlignment.Bottom, EAxisAlignment.Top].includes(alignment)) {
        return true;
    }
    if ([EAxisAlignment.Right, EAxisAlignment.Left].includes(alignment)) {
        return false;
    }
    return undefined;
};
exports.getIsVertical = (alignment) => {
    if (exports.getIsHorizontal(alignment) === false) {
        return true;
    }
    if (exports.getIsHorizontal(alignment) === true) {
        return false;
    }
    return undefined;
};
