// scichart/types/NumericFormat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENumericFormat = void 0;
/**
 * Enumeration constants for standard formatting strings
 */
var ENumericFormat;
(function (ENumericFormat) {
    /**
     * No format, return the string representation unchanged
     */
    ENumericFormat["NoFormat"] = "NoFormat";
    /**
     * Format as an integer
     */
    ENumericFormat["Decimal_0"] = "Decimal_0";
    /**
     * Format as a floating-point number with 1 decimal place
     */
    ENumericFormat["Decimal_1"] = "Decimal_1";
    /**
     * Format as a floating-point number with 2 decimal places
     */
    ENumericFormat["Decimal_2"] = "Decimal_2";
    /**
     * Format as a floating-point number with 4 decimal places
     */
    ENumericFormat["Decimal_4"] = "Decimal_4";
    /**
     * Format as a date in format DDMMYYYY
     */
    ENumericFormat["Date_DDMMYYYY"] = "Date_DDMMYYYY";
    /**
     * Format as a date in format DDMM
     */
    ENumericFormat["Date_DDMM"] = "Date_DDMM";
    /**
     * Format as a date in format HHMM
     */
    ENumericFormat["Date_HHMM"] = "Date_HHMM";
})(ENumericFormat = exports.ENumericFormat || (exports.ENumericFormat = {}));
