// scichart/types/SceneEntityType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ESceneEntityType = void 0;
/**
 * Enumeration constants to define the type of {@link BaseSceneEntity3D | Scene Entity }
 */
var ESceneEntityType;
(function (ESceneEntityType) {
    /**
     * The entity is an axis cube
     */
    ESceneEntityType[ESceneEntityType["AxisCubeEntity"] = 0] = "AxisCubeEntity";
    /**
     * The entity is an Xyz gizmo
     */
    ESceneEntityType[ESceneEntityType["GizmoEntity"] = 1] = "GizmoEntity";
    /**
     * The entity is the root scene entity
     */
    ESceneEntityType[ESceneEntityType["RootSceneEntity"] = 2] = "RootSceneEntity";
    /**
     * The entity is a Scatter 3D or Bubble 3D entity
     */
    ESceneEntityType[ESceneEntityType["ScatterPointsSceneEntity"] = 3] = "ScatterPointsSceneEntity";
    /**
     * A generic WebAssembly native entity
     */
    ESceneEntityType[ESceneEntityType["SCRTSceneEntity"] = 4] = "SCRTSceneEntity";
    /**
     * The entity is a Surface Mesh or Surface Plot 3D entity
     */
    ESceneEntityType[ESceneEntityType["SurfaceMeshSceneEntity"] = 5] = "SurfaceMeshSceneEntity";
})(ESceneEntityType = exports.ESceneEntityType || (exports.ESceneEntityType = {}));
