// scichart/types/SeriesType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ESeriesType = void 0;
/**
 * Enumeration constants to define the type of series, or chart type that a {@link BaseRenderableSeries} represents
 */
var ESeriesType;
(function (ESeriesType) {
    ESeriesType["BandSeries"] = "Band Series";
    ESeriesType["SplineBandSeries"] = "Spline Band Series";
    ESeriesType["BubbleSeries"] = "Bubble Series";
    ESeriesType["CandlestickSeries"] = "Candlestick Series";
    ESeriesType["ColumnSeries"] = "Column Series";
    ESeriesType["LineSeries"] = "Line Series";
    ESeriesType["SplineLineSeries"] = "Spline Line Series";
    ESeriesType["MountainSeries"] = "Mountain Series";
    ESeriesType["SplineMountainSeries"] = "Spline Mountain Series";
    ESeriesType["OhlcSeries"] = "OHLC Series";
    ESeriesType["StackedColumnSeries"] = "Stacked Column Renderable Series";
    ESeriesType["StackedMountainSeries"] = "Stacked Mountain Renderable Series";
    ESeriesType["UniformContoursSeries"] = "Uniform Contours Series";
    ESeriesType["UniformHeatmapSeries"] = "Uniform Heatmap Series";
    ESeriesType["VerticallyStackedMountainCollection"] = "Vertically Stacked Mountain Collection";
    ESeriesType["VerticallyStackedColumnCollection"] = "Vertically Stacked Column Collection";
    ESeriesType["ScatterSeries"] = "XY Scatter Series";
})(ESeriesType = exports.ESeriesType || (exports.ESeriesType = {}));
