// scichart/Charting3D/Visuals/Primitives/RenderableSeriesSceneEntityState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderableSeriesSceneEntityState = void 0;
/**
 * State object for a 3D {@link RenderableSeriesSceneEntity} that contains a set of boolean flags and other state types
 * (e.g. TAxisCubeState) that determine whether underlying graphics resources of the Scene Entity are up-to-date and reflect
 * the current state of its properties and the date it visualizes.
 */
class RenderableSeriesSceneEntityState {
    constructor(isInitialStateProperty = true, isDataSeriesModifiedProperty = true, isRenderableSeriesPropertyChangedProperty = true, isAxisCubeStateChangedProperty = true, axisCubeState = {
        xVisibleMin: 0,
        xVisibleMax: 0,
        yVisibleMin: 0,
        yVisibleMax: 0,
        zVisibleMin: 0,
        zVisibleMax: 0,
        xWorldDimension: 0,
        yWorldDimension: 0,
        zWorldDimension: 0
    }) {
        this.isInitialStateProperty = isInitialStateProperty;
        this.isDataSeriesModifiedProperty = isDataSeriesModifiedProperty;
        this.isRenderableSeriesPropertyChangedProperty = isRenderableSeriesPropertyChangedProperty;
        this.isAxisCubeStateChangedProperty = isAxisCubeStateChangedProperty;
        this.axisCubeState = axisCubeState;
    }
    /**
     * Gets whether the Scene Entity is in the initial state, meaning all the underlying resources must be (re-)created.
     */
    get isInitialState() {
        return this.isInitialStateProperty;
    }
    /**
     * Sets whether the Scene Entity is in the initial state, meaning all the underlying resources must be (re-)created.
     */
    setInitialState() {
        this.isInitialStateProperty = true;
    }
    /**
     * Gets whether the Data Series has been modified since last update.
     */
    get isDataSeriesModified() {
        return this.isDataSeriesModifiedProperty;
    }
    /**
     * Gets whether a property of the Renderable Series has been changed since last update.
     */
    get isRenderableSeriesPropertyChanged() {
        return this.isRenderableSeriesPropertyChangedProperty;
    }
    /**
     * Sets whether a property of the Renderable Series has been changed.
     */
    setRenderableSeriesPropertyChanged() {
        this.isRenderableSeriesPropertyChangedProperty = true;
    }
    /**
     * Gets whether an Axis Cube state has been changed since last update.
     */
    get isAxisCubeStateChanged() {
        return this.isAxisCubeStateChangedProperty;
    }
    /**
     * Checks if a state of the Renderable Series and a state of other objects in the Render Pass (e.g. Axis Cube)
     * has been changed. Sets corresponding flags, if any.
     * @param rs The Rendrable Series to check
     * @param rpi The current {@link RenderPassInfo3D} to check
     */
    validate(rs, rpi) {
        this.isDataSeriesModifiedProperty = rs.dataSeries.isModified;
        const fpTolerance = 0.001;
        this.isAxisCubeStateChangedProperty =
            Math.abs(this.axisCubeState.xVisibleMin - rpi.xCalc.visibleMin) > fpTolerance ||
                Math.abs(this.axisCubeState.xVisibleMax - rpi.xCalc.visibleMax) > fpTolerance ||
                Math.abs(this.axisCubeState.yVisibleMin - rpi.yCalc.visibleMin) > fpTolerance ||
                Math.abs(this.axisCubeState.yVisibleMax - rpi.yCalc.visibleMax) > fpTolerance ||
                Math.abs(this.axisCubeState.zVisibleMin - rpi.zCalc.visibleMin) > fpTolerance ||
                Math.abs(this.axisCubeState.zVisibleMax - rpi.zCalc.visibleMax) > fpTolerance ||
                Math.abs(this.axisCubeState.xWorldDimension - rpi.worldDimensions.x) > fpTolerance ||
                Math.abs(this.axisCubeState.yWorldDimension - rpi.worldDimensions.y) > fpTolerance ||
                Math.abs(this.axisCubeState.zWorldDimension - rpi.worldDimensions.z) > fpTolerance;
        return (!this.isInitialState &&
            !this.isDataSeriesModified &&
            !this.isDataSeriesModified &&
            !this.isAxisCubeStateChanged);
    }
    /**
     * Resets all the state changed flags and saves the current state of Rederable Series
     * and state of other objects in the Render Pass (e.g. Axis Cube).
     * @param rs The Rendrable Series, whose state gets save
     * @param rpi The current {@link RenderPassInfo3D}, whose state gets save
     */
    reset(rs, rpi) {
        // Reset all flags
        this.isInitialStateProperty = false;
        this.isDataSeriesModifiedProperty = false;
        this.isRenderableSeriesPropertyChangedProperty = false;
        this.isAxisCubeStateChangedProperty = false;
        // Update Axis Cube State
        this.axisCubeState = {
            xVisibleMin: rpi.xCalc.visibleMin,
            xVisibleMax: rpi.xCalc.visibleMax,
            yVisibleMin: rpi.yCalc.visibleMin,
            yVisibleMax: rpi.yCalc.visibleMax,
            zVisibleMin: rpi.zCalc.visibleMin,
            zVisibleMax: rpi.zCalc.visibleMax,
            xWorldDimension: rpi.worldDimensions.x,
            yWorldDimension: rpi.worldDimensions.y,
            zWorldDimension: rpi.worldDimensions.z
        };
        // Reset isModified flag of the Data Series
        // TODO: consider a better place for that
        rs.dataSeries.resetModified();
    }
}
exports.RenderableSeriesSceneEntityState = RenderableSeriesSceneEntityState;
